### PR TITLE
[DataGrid] Improve row selection propagation trigger

### LINKED ISSUE #15045: Filtering with `isRowSelectable` and `treeData` results in a crash when a row is selected
### Steps to reproduce

Link to live example: https://stackblitz.com/edit/react-ayw1zv-qxwdvx?file=Demo.js

Steps:
1. Have a DataGrid with `treeData={true}` and `isRowSelectable={() => true}`. 
2. Select a row, for example the first.
3. Filter the rows with one of the header filters, for example "Commodity". Make sure to filter in a way that the selected row would NOT be shown.

### Current behavior

The page crashes with the following error: 

```
Error in /turbo_modules/@mui/x-data-grid@7.21.0/node/hooks/features/rows/useGridParamsApi.js (33:13)
No row with id #auto-generated-row-null/4bfc3018-3d32-513e-9913-5b81a56718a0 found
```

### Expected behavior

The page should not crash and filter correctly.

### Context

I'm aware that the error ```MUI X: The `filterMode="server"` prop is not available when the `treeData` is enabled.``` is logged. When not using `isRowSelectable` everything works fine though.

I want to select child rows in the DataGrid and keep that selection when using filters. 

Currently, I bypass the issues by not setting `isRowSelectable`, but by providing a custom `renderCell` to my custom checkbox column that is mostly based on `GRID_CHECKBOX_SELECTION_COL_DEF`. 

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  System:
    OS: macOS 14.7
  Binaries:
    Node: 22.8.0 - ~/.nvm/versions/node/v22.8.0/bin/node
    npm: 10.8.2 - ~/.nvm/versions/node/v22.8.0/bin/npm
    pnpm: Not Found
  Browsers:
    Chrome: 130.0.6723.59
    Edge: Not Found
    Safari: 17.6
  npmPackages:
    @emotion/react: 11.13.3 => 11.13.3 
    @emotion/styled: 11.13.0 => 11.13.0 
    @mui/base:  5.0.0-beta.40 
    @mui/core-downloads-tracker:  5.16.7 
    @mui/icons-material: 6.1.4 => 6.1.4 
    @mui/joy: 5.0.0-beta.48 => 5.0.0-beta.48 
    @mui/material: 6.1.4 => 6.1.4 
    @mui/private-theming:  6.1.4 
    @mui/styled-engine:  6.1.4 
    @mui/system:  6.1.4 
    @mui/types:  7.2.18 
    @mui/utils:  5.16.6 
    @mui/x-data-grid: 7.21.0 => 7.21.0 
    @mui/x-data-grid-premium: 7.21.0 => 7.21.0 
    @mui/x-data-grid-pro:  7.21.0 
    @mui/x-date-pickers: 7.21.0 => 7.21.0 
    @mui/x-internals:  7.21.0 
    @mui/x-license:  7.21.0 
    @types/react: 18.3.11 => 18.3.11 
    react: 18.3.1 => 18.3.1 
    react-dom: 18.3.1 => 18.3.1 
    typescript:  5.5.4 
```
</details>

**Search keywords**: DataGrid isRowSelectable treeData filter
**Order ID**: 95669

### LINKED ISSUE #15068: [data grid] `onRowSelectionModelChange` is called when `treeData` is enabled and `rows` changes
### Steps to reproduce

Link to live example: https://codesandbox.io/p/sandbox/hopeful-maxwell-dxvvlw?from-embed=&workspaceId=67e261df-ad7c-4156-98c4-4a0336fb54f4

Steps:
1. Open the console and click the button above the grid that says "add a row"

### Current behavior

Despite the row selection not changing, `onRowSelectionModelChange` is being called. See the console output.

### Expected behavior

`onRowSelectionModelChange` should not be called when changing rows, unless the row selection changes too. In the example, there should be no console output.

If it is technically impossible not to call `onRowSelectionModelChange` in this case, it should at least return the same array (not a new one with the same values) so I can easily check whether the row selection model changed.

### Context

This worked correctly in v7.18.0. It breaks in v7.19.0 and hasn't been fixed since.

We reset parts of our UI when the row selection changes. For this, we use `onRowSelectionModelChange`. We cannot compare the row selection array content with the previous content because of performance. There are too many selected rows.

This only happens when `treeData` is enabled.

### Your environment

<details>
  <summary><code>bunx @mui/envinfo</code></summary>

```
  System:
    OS: Linux 5.15 Ubuntu 22.04.3 LTS 22.04.3 LTS (Jammy Jellyfish)
  Binaries:
    Node: 20.11.1 - ~/.nvm/versions/node/v20.11.1/bin/node
    npm: 10.2.4 - ~/.nvm/versions/node/v20.11.1/bin/npm
    pnpm: 9.6.0 - /mnt/c/Users/lgoldstein/AppData/Roaming/npm/pnpm
  Browsers:
    Chrome: 121.0.6167.85
  npmPackages:
    @emotion/react: ^11.11.4 => 11.11.4 
    @emotion/styled: ^11.11.5 => 11.11.5 
    @mui/base:  5.0.0-beta.42 
    @mui/core-downloads-tracker:  6.1.2 
    @mui/icons-material: ^6.1.2 => 6.1.2 
    @mui/lab: ^6.0.0-beta.10 => 6.0.0-dev.240424162023-9968b4889d 
    @mui/material: ^6.1.2 => 6.1.2 
    @mui/private-theming:  6.1.0 
    @mui/styled-engine:  6.1.0 
    @mui/system:  6.1.0 
    @mui/types:  7.2.15 
    @mui/utils:  6.1.0 
    @mui/x-data-grid:  7.21.0 
    @mui/x-data-grid-premium: ^7.21.0 => 7.21.0 
    @mui/x-data-grid-pro:  7.21.0 
    @mui/x-date-pickers:  7.21.0 
    @mui/x-date-pickers-pro: ^7.21.0 => 7.21.0 
    @mui/x-internals:  7.21.0 
    @mui/x-license: ^7.18.0 => 7.18.0 
    @types/react: ~18.3.11 => 18.3.11 
    react: ^18.3.1 => 18.3.1 
    react-dom: ^18.3.1 => 18.3.1 
    typescript: ^5.6.3 => 5.6.3 
```
</details>

Reproducible in Chrome and Firefox.

**Search keywords**: onRowSelectionModelChange, treeData, rows, reference changes
**Order ID**: 95800

### LINKED ISSUE #15079: [data grid] `onRowSelectionModelChange` firing twice
### Search keywords

onRowSelectionModelChange

### Latest version

- [X] I have tested the latest version

### Affected products

Data Grid

### Steps to reproduce

Link to live example: (required)
https://codesandbox.io/p/sandbox/headless-microservice-x38xc6

Steps:
1. Use  getRowId={(row) => row.id}, instead of defining id field in the column definition
2. set filterMode="server"

### Current behavior

The `onRowSelectionModelChange` handler is triggered twice when trying to update any state within the handler, if `getRowId` is used to set the unique row ID and `filterMode` is set to 'server'. However, when `filterMode` is set to 'client', the handler is only triggered once.

### Expected behavior

`onRowSelectionModelChange` should be triggered only once.

### Context

The response data doesn't include an id field but uses a different key as the record identifier. I am using getRowId to assign the id field.

### Your environment

System:
    OS: Linux 6.8 Ubuntu 22.04.5 LTS 22.04.5 LTS (Jammy Jellyfish)
  Binaries:
    Node: 22.1.0 - ~/.nvm/versions/node/v22.1.0/bin/node
    npm: 10.7.0 - ~/.nvm/versions/node/v22.1.0/bin/npm
    pnpm: Not Found
  Browsers:
    Chrome: 130.0.6723.58
  npmPackages:
    @emotion/react: ^11.13.3 => 11.13.3 
    @emotion/styled: ^11.13.0 => 11.13.0 
    @mui/core-downloads-tracker:  6.1.5 
    @mui/material: ^6.1.5 => 6.1.5 
    @mui/private-theming:  6.1.5 
    @mui/styled-engine:  6.1.5 
    @mui/system:  6.1.5 
    @mui/types:  7.2.18 
    @mui/utils:  6.1.5 
    @mui/x-data-grid: ^7.21.0 => 7.21.0 
    @mui/x-internals:  7.21.0 
    @types/react: ^18.3.10 => 18.3.10 
    react: ^18.3.1 => 18.3.1 
    react-dom: ^18.3.1 => 18.3.1 
    typescript: ^4.9.5 => 4.9.5 

### Order ID or Support key 💳 (optional)

_No response_

### LINKED ISSUE #15217: [data grid] using treeData and onRowSelectionModelChange with controlled rowSelectionModel causing infinite render loop
### Steps to reproduce

Link to live example: [(required)](https://codesandbox.io/p/sandbox/dgpro-latest-row-selection-bug-forked-57mkp3?file=%2Fsrc%2FDemo.tsx%3A120%2C11-120%2C31)

Steps:
1. Click the row selection checkbox on Sarah, Thomas, or Mary

Here's a video of the behavior:
https://github.com/user-attachments/assets/253ec8c4-b2ed-4154-a3af-4c5d59bea555

### Current behavior

In the case when we're using `treeData` and `onRowSelectionModelChange` with controlled `rowSelectionModel` - the selected row toggles between selected and unselected when the row selection checkbox is used. 

### Expected behavior

In the case when we're using `treeData` and `onRowSelectionModelChange` with controlled `rowSelectionModel` - the selected row reflects it's selected state when remains selected when using the row selection checkbox.

### Context

When you click a row to select row selection, the component goes into an infinite loop of calling `onRowSelectionModelChange`. The first time it calls it, it contains the id of the row you just selected and then it is called again with an empty list in the case of this example, causing the flashing behavior. In general, `onRowSelectionModelChange` gets called again with whatever the `rowSelectionModel` was before the click occurred. For example, if there's an initially selected row when it first mounts - it will go back to that. 

It appears as though this issue was introduced in 7.19.0. 

I'm not sure if it's related to other `onRowSelectionModelChange` open issues. This could potentially be related: https://github.com/mui/mui-x/issues/14859

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
System:
    OS: macOS 13.6.2
  Binaries:
    Node: 20.8.1 - ~/.asdf/installs/nodejs/20.8.1/bin/node
    npm: 10.1.0 - ~/.asdf/plugins/nodejs/shims/npm
    pnpm: Not Found
  Browsers:
    Chrome: 130.0.6723.71
    Edge: Not Found
    Safari: 17.1

```
</details>

**Search keywords**: data grid premium pro treeData onRowSelectionModelChange rowSelectionModel rowSelection
**Order ID**: 82849

### PR DESCRIPTION
**Pull request by MBilalShafi**
Fixes #15068 (Pro plan)
- Before: https://codesandbox.io/p/sandbox/hopeful-maxwell-dxvvlw
- After: https://codesandbox.io/p/sandbox/mui-mui-x-x-data-grid-forked-ssxqym

Fixes #15045 (Pro plan)
- Before: https://stackblitz.com/edit/react-ayw1zv-qxwdvx?file=Demo.js
- After: https://codesandbox.io/p/sandbox/mui-mui-x-x-data-grid-forked-2hkqsq

Fixes #15079 (Community plan)
- Before: https://codesandbox.io/p/sandbox/headless-microservice-x38xc6
- After: https://codesandbox.io/p/sandbox/mui-mui-x-x-data-grid-forked-zr6gzz

Fixes #15217 (Pro plan)
- Before: https://codesandbox.io/p/sandbox/dgpro-latest-row-selection-bug-forked-57mkp3
- After: https://codesandbox.io/p/sandbox/mui-mui-x-x-data-grid-forked-3j6y7t

### Follow-up todos:

- [ ] Add test coverage - will do so as a follow-up PR, there are multiple corner cases to test and it'd take some digging

### CODE REVIEW

File: packages/x-data-grid/src/hooks/features/rowSelection/useGridRowSelection.ts [hunk 1]
```diff
@@ -449,14 +461,22 @@ export const useGridRowSelection = (
...
       const currentSelection = gridRowSelectionStateSelector(apiRef.current.state);
+      const rowsLookup = gridRowsLookupSelector(apiRef);
       const filteredRowsLookup = gridFilteredRowsLookupSelector(apiRef);
 
       // We clone the existing object to avoid mutating the same object returned by the selector to others part of the project
       const selectionLookup = { ...selectedIdsLookupSelector(apiRef) };
 
+      const isNonExistent = (id: GridRowId) => {
+        if (props.filterMode === 'server') {
+          return !rowsLookup[id];
```
Comment (MBilalShafi) [inline packages/x-data-grid/src/hooks/features/rowSelection/useGridRowSelection.ts:472]: There is no `filteredRowsLookup` when `filterMode='server'`

**Comment by shubhsk88** [comment]
Is there any idea when this PR can be merged? This seems to keep happening to me whenever I try to select the checkbox in parent/descendant, and then it causes the table to rerender infinitely. This is blocking our workflow. Would love to get an estimation around this

**Comment by KenanYusuf** [comment]
@MBilalShafi is it expected that in [this demo](https://codesandbox.io/p/sandbox/mui-mui-x-x-data-grid-forked-3j6y7t), selecting the child rows cause the group to collapse?

https://github.com/user-attachments/assets/393e302e-ad39-4e1b-8924-a95399919f88

**Comment by MBilalShafi** [comment]
@KenanYusuf It's a problem with the demo itself and not the Data Grid, here's some context on why it happens: https://github.com/mui/mui-x/issues/7771#issuecomment-1410218395

Here's a fixed version: https://codesandbox.io/p/sandbox/mui-mui-x-x-data-grid-forked-cgpss2

**Review by KenanYusuf — APPROVED** [review]

**Comment by oliviertassinari** [comment]
It looks like the CI didn't run here and broke master:

SCR-20241105-lktd

I noticed this on #15271.
